File: src/main/java/slowraytracer/Pixmap.java
package slowraytracer;

import java.awt.image.BufferedImage;
import java.util.Objects;

public final class Pixmap {

    private final int width;
    private final int height;
    private final int[] data;

    public Pixmap(final int width, final int height) {
        this.width = Validations.requirePositive(width, "width");
        this.height = Validations.requirePositive(height, "height");
        data = new int[width * height];
    }

    public void set(final int x, final int y, final int color) {
        Validations.requireRange(x, 0, width, "X-coordinate");
        Validations.requireRange(y, 0, height, "Y-coordinate");
        data[x + y * width] = color;
    }

    public void fill(final Mapper mapper) {
        Objects.requireNonNull(mapper);
        for (var y = 0; y < height; y++) {
            for (var x = 0; x < width; x++) {
                data[x + y * width] = mapper.map(x, y);
            }
        }
    }

    public int get(final int x, final int y) {
        Validations.requireRange(x, 0, width, "X-coordinate");
        Validations.requireRange(y, 0, height, "Y-coordinate");
        return data[x + y * width];
    }

    public BufferedImage asBufferedImage() {
        final BufferedImage bufferedImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
        bufferedImage.setRGB(0, 0, width, height, data, 0, width);
        return bufferedImage;
    }

    public int width() {
        return width;
    }

    public int height() {
        return height;
    }

    @FunctionalInterface
    public interface Mapper {

        int map(int x, int y);
    }
}

File: src/main/scala/slowraytracer/Pixmap.scala
package slowraytracer

import java.awt.image.BufferedImage

class Pixmap(width: Int, height: Int) {

  Validations.assertPositive(width, "width")
  Validations.assertPositive(height, "height")

  private val data: Array[Color] = new Array[Color](width * height)

  def width(): Int = width

  def height(): Int = height

  def set(x: Int, y: Int, color: Color): Unit = {
    Validations.assertRange(x, 0, width, "X-coordinate")
    Validations.assertRange(y, 0, height, "Y-coordinate")
    data(x + y * width) = color;
  }

  def fill(mapper: (Int, Int) => Color): Unit = {
    for (y <- 0 until height) {
      for (x <- 0 until width) {
        data(x + y * width) = mapper(x, y)
      }
    }
  }

  def get(x: Int, y: Int): Color = {
    Validations.assertRange(x, 0, width, "X-coordinate")
    Validations.assertRange(y, 0, height, "Y-coordinate")
    data(x + y * width)
  }

  def asBufferedImage: BufferedImage = {
    val bufferedImage: BufferedImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
    bufferedImage.setRGB(0, 0, width, height, data.map(_.argb), 0, width)
    bufferedImage
  }
}
